### DISCUSSION TITLE
GC6609 stepper drivers instead of original TMC2208 being used at certain batches

### DISCUSSION THREAD

**Discussion by 1coderookie** [1coderookie/KobraGoNeoInsights#13] [General]
Thanks to @aspiringnobody for discovering and reaching out with his finding that there are also batches of the mainboard which come with "GC6609" stepper drivers instead of original TMC2208 ones. According to @aspriningnobody these ones are allegedly clones of TMC2209 drivers.

**Comment by aspiringnobody** [comment]
There is very little information about the GC6609 aka S6609.  It seems to be made by FYSETC (at least the S6609 variant of it is) but probably because it infringes on copyright they don't publicize it at all.  There isn't a datasheet that I can find.  Someone did a writeup of these in module form -- this is the "best" information I can find about them, and it's not much.

[https://hackaday.io/project/184230-printer-repair-gd0137/log/204221-dp2-trying-the-gc6609-stepper-drivers](https://hackaday.io/project/184230-printer-repair-gd0137/log/204221-dp2-trying-the-gc6609-stepper-drivers)

Allegedly they are TMC2209 clones, but noone has ever tested to see if they include the more advanced TMC2209 features like sensorless homing and working with linear advance.  Some people think they are a TMC2208 clone with current capabilities similar to the TMC2209 so they are marketed as 2209 clones even though they are, in fact, beefed up 2208 clones.

I've replaced them on my motherboard with real TMC2209s so I can use LA.  Experiments are being done to see if the GC6609, when marlin is compiled with 
```
#define X_DRIVER_TYPE  TMC2209
#define Y_DRIVER_TYPE  TMC2209
#define Z_DRIVER_TYPE  TMC2209
#define E0_DRIVER_TYPE TMC2209
```
is able to use LA without the marlin bugfix.  These drivers can't use the marlin bugfix, near as I can tell there is an issue with missed steps when the driver handles high acceleration.  Forcing SpreadCycle didn't make any difference.  They might be able to handle it on the "stock" jokubasver firmware if they are really 2209 clones.

My real TMC2209s handle LA just fine on the @jokubasver firmware without @ElijahCuff 's marlin bugfix branch.

Sadly I can't do any of this testing right now because my board has been modified.  I've got another board on order, and hopefully since it's new from AC it will come with the clone drivers and I can do some more testing.  Will update here when I know more.

Klipper shouldn't care -- TMC2208s and TMC2209s are functionally identical.  I would continue to tell Klipper it is a TMC2208 unless we learn more.  The TMC2209 is pin and UART compatible with the TMC2208 so it shouldn't matter.  You just won't have access to sensorless homing.

**Reply by 1coderookie** [reply]
Thanks!  

~~What I'm wondering right now though: does this has any impact on Klipper (as the cfg files there also have the 2208 set)? Do you know anything about that?~~ (you just answered it already, thanks ;) )
I'm not using the stock fw and don't use the Marlin based fw versions of other users who activated certain functions, so does this maybe also have any impact on the regular stock fw (just the plain stock fw, not modded in any way)?

Btw: is this from a Go or a Neo?  
And how old is the machine?  
I wonder if AC maybe recently changed the drivers to these GC ones - they also changed the part cooling fan of the Neo (which doesn't work with PWM anymore now..) somewhere around the middle of this year probably (dunno when exactly), so I wouldn't be surprised actually if they now also changed the stepper drivers due to lower costs or whatever..  

~~(Sidenote: you wrote "Experiments are being done to see if the GC2209,(...)" but you meant GC6609, right?)~~

**Reply by aspiringnobody** [reply]
Fixed.  I added an edit above about Klipper but I think for now it won't make any difference and it's best to continue to tell Klipper they are TMC2208s.  There isn't anything to be gained except sensorless homing which would be of dubious accuracy on the go/neo anyway because of how weak the frame is.

**Reply by 1coderookie** [reply]
Thanks, saw it and edited my post as well ;)  
Is there any way to get the info from the mobo directly, like a certain g-code command or whatever, instead of having to pull off the heatsink to see which kind is actually being used?

**Reply by aspiringnobody** [reply]
This is from a relatively new Neo.  Early summer purchase from Amazon (US).  There should be no issues with the stock firmware, Anycubic continues to use 2208 settings in Marlin.  I've noticed that these steppers don't behave well with the transition between SpreadCycle and StealthChop2 that happens when #define HYBRID_THRESHOLD is uncommented -- so I'd leave that feature off -- but it is off by default anyway.

**Reply by jokubasver** [reply]
I don't think that the Trigorilla board has traces and accompanying resistors connecting the TMC UART pin to the MCU, so TMC configuration won't work in that case.

**Reply by CloudedQuartz** [reply]
hi @aspiringnobody , i was wondering, did you manage to get the uart slave addressing working with genuine tmc2209s? like the r65 r66 mod on the standard kobra?

**Comment by aspiringnobody** [comment]
> Thanks, saw it and edited my post as well ;) Is there any way to get the info from the mobo directly, like a certain g-code command or whatever, instead of having to pull off the heatsink to see which kind is actually being used?  

Nope, I don't think this is possible. M122 doesn't seem to tell you what driver it is talking to and even if it did the clones would say they were genuine TMC drivers. Marlin is compiled by default to think they are TMC2208s

**Reply by 1coderookie** [reply]
Dammit, that would've been real handy to find out which one is being used.. :( ;)

**Reply by aspiringnobody** [reply]
Super long shot, but my board has silver heatsinks and I noticed the pictures you include have black ones.  It's possible that they switched heatsinks at the same time.

I'd say it's more likely the heatsinks change all the time as prices fluctuate.

**Reply by 1coderookie** [reply]
Yes, mine have black ones! Thanks, that might be a good indicator tho as they maybe just cut the costs overall. I wonder what else might have been changed we aren't aware of at this point..

**Comment by 1coderookie** [comment]
I added a note to the mobo chapter as an expandable textblock (above the first picture of the mobo) - maybe check it out if there's anything I should probably add at that point..

**Reply by aspiringnobody** [reply]
Looks good.  Now we need to wait for my replacement board to come and hope it has the new chips on it so I can load my TMC2209 firmware on it and see if LA works.  It would be nice if these actually ended up being an "upgraded" board.

**Reply by 1coderookie** [reply]
Haha, yeah, that would be cool :D  
Looking forward to your findings!
Thx again!

**Comment by aspiringnobody** [comment]
> I don't think that the Trigorilla board has traces and accompanying resistors connecting the TMC UART pin to the MCU, so TMC configuration won't work in that case.

If UART isn’t wired on the board shouldn’t we be using TMC2208_STANDALONE?

**Reply by jokubasver** [reply]
Actually, seems like it is wired, but only one UART pin for 4 drivers, so TMC2208 UART is out of the question, as described in this discussion here: https://www.reddit.com/r/anycubic/comments/10qopuj/comment/jikhsiu/?context=3
So in theory - TMC2209 would work then.

As for the GC6609 - I remember seeing references to TMC6609 in the Marlin source code, also mentioned here: https://gist.github.com/SteveGotthardt/2ed249f197f5928eaf5c3d3877d55503?permalink_comment_id=4454841#gistcomment-4454841

My Kobra Neo was ordered at around New Years, and I haven't taken a look at the mainboard yet, so no idea what drivers are soldered in my unit. Since TMC2209 drivers are pin-to-pin compatible, I might try soldering them in and seeing if UART works in Klipper.

**Reply by aspiringnobody** [reply]
With Marlin compiled TMC2209 I’m able to toggle to spreadcycle and use the hybrid mode.  So I assume that means UART works fine with the 2209s. Then again I could do that with the GC6609 drivers as well… but when those were on the board I was using your firmware with TMC2208.  I’ve not seen any Neo/Go firmware compiled using TMC2208_STANDALONE.

I do know Anycubic do use that mode because that’s how my Vyper was until did the hardware mod to enable UART.  But that has 2209s.

**Reply by jokubasver** [reply]
Did some digging - TMC UART won't work with Klipper - the Anycubic Marlin fork uses SoftwareSerial to communicate with the steppers in UART - not a hardware serial interface - and since Klipper does not support software serial, the only way it would work is with a hardware modification - rewiring the TMC UART pin to be connected to some other serial port (for example the LCD)

**Comment by aspiringnobody** [comment]
> Did some digging - TMC UART won't work with Klipper - the Anycubic Marlin fork uses SoftwareSerial to communicate with the steppers in UART - not a hardware serial interface - and since Klipper does not support software serial, the only way it would work is with a hardware modification - rewiring the TMC UART pin to be connected to some other serial port (for example the LCD)

So M122 on my modified board with TMC2209s fails to connect to X Y and Z -- but connects to E.  M122 on my board with GC6609s doesn't connect to any of them.  On the stock board, the pots to set vref are used by the stepper for X and Y (verified by turning them down and seeing how hard it is to make the steppers skip).

Is the anycubic firmware bad from the get-go?  Should we be using TMC2208_STANDALONE for X Y and Z, and TMC2208 for E?  I'm wondering if Anycubic bothered to configure the MS1/MS2 resistors for the slave addresses to enable TMC2209 UART mode.

**Reply by jokubasver** [reply]
Yeah, it's bad, it's a mish mash on both the firmware and the hardwire (bad pinout labeling on the LCD connector on the mainboard for example)

**Reply by aspiringnobody** [reply]
I just flashed TMC2209_STANDALONE for X Y Z E on my modded motherboard, and nothing exploded.  Although I didn't expect anything would.

Is there an easy way to verify what blank resistor pads are used for MS1/MS2 without desoldering the drivers to get access to the pads?  I'm not super enthused about having to remove the heatsinks again since I used real thermal glue instead of thermal plaster (it's what I had on hand).  Under each driver are two empty pads next to each other on the left -- I'm guessing these are good canidates.  Assuming these are pull down resistors that would mean that they are all configured for 16uSteps and also by extension slave address 3 -- so any one of them could be responding when trying UART -- but slave address 3 is set in the anycubic firmware to E so that would explain why that one worked...

It would be easy enough to use some pencil to make some temp connections to verify if I knew it wasn't going to blow anything up.

**Reply by robotm-team** [reply]
Hi.
My printer is a KobraNEO.
I am not an advanced user, I am completely green when it comes to 3d printing.
For a klipper for me it is too early (which I sincerely regret) because I am aware of my unfamiliarity with the subject.
To the point:
I got an extra MB from tech support (I don't know why because I advertised a crooked hotbed...) and out of curiosity, after reading this topic I took the heatsinks off the controllers. It turned out that I just have these fake GC6609s.
![mobo](https://github.com/1coderookie/KobraGoNeoInsights/assets/76626315/5128574d-7d97-4577-a21f-9261b803fee0)

![box](https://github.com/1coderookie/KobraGoNeoInsights/assets/76626315/8585ed3a-d199-4689-80e3-bbc9a2aa8d4b)

@aspiringnobody wrote that he replaced these clones on his board with the original TMC2209.
My question is:
Is it worth "playing" with such modifications?
What are the real benefits of this procedure (speed, stability, etc.)?
If I wanted to risk it, would I rather buy the whole controller and scalp it with the chips or maybe buy on e.g. Aliexpress the chips themselves?

I currently have the firmware @ElijahCuff uploaded to the motherboard which is installed marked "Linear Advance Beta Latest - released this Sep 19" and calling M122 shows this log:

```
Send: M122
Recv: 		X	Y	Z	E
Recv: Address		0	1	2	3
Recv: Enabled		false	false	false	false
Recv: Set current	800	800	800	600
Recv: RMS current	1436	1436	1436	1049
Recv: MAX current	2025	2025	2025	1479
Recv: Run current	25/31	25/31	25/31	18/31
Recv: Hold current	12/31	12/31	12/31	9/31
Recv: CS actual	0/31	0/31	0/31	0/31
Recv: PWM scale
Recv: vsense		0=.325	0=.325	0=.325	0=.325
Recv: stealthChop	false	false	false	false
Recv: msteps		256	256	256	256
Recv: interp		false	false	false	false
Recv: tstep		0	0	0	0
Recv: PWM thresh.
Recv: [mm/s]
Recv: OT prewarn	false	false	false	false
Recv: pwm scale sum	0	0	0	0
Recv: pwm scale auto	0	0	0	0
Recv: pwm offset auto	0	0	0	0
Recv: pwm grad auto	0	0	0	0
Recv: off time	0	0	0	0
Recv: blank time	16	16	16	16
Recv: hysteresis
Recv:  -end		-3	-3	-3	-3
Recv:  -start		1	1	1	1
Recv: Stallguard thrs	0	0	0	0
Recv: uStep count	0	0	0	0
Recv: DRVSTATUS	X	Y	Z	E
Recv: sg_result	0	0	0	0
Recv: stst		*	*	*	*
Recv: olb
Recv: ola
Recv: s2gb
Recv: s2ga
Recv: otpw
Recv: ot
Recv: 157C
Recv: 150C
Recv: 143C
Recv: 120C
Recv: s2vsa
Recv: s2vsb
Recv: Driver registers:
Recv: 		X	0x00:00:00:00	 Bad response!
Recv: 		Y	0x00:00:00:00	 Bad response!
Recv: 		Z	0x00:00:00:00	 Bad response!
Recv: 		E	0x00:00:00:00	 Bad response!
Recv: 
Recv: 
Recv: Testing X connection... Error: All LOW
Recv: Testing Y connection... Error: All LOW
Recv: Testing Z connection... Error: All LOW
Recv: Testing E connection... Error: All LOW
Recv: ok
```

I don't know if this helps with anything but if it would tell anyone anything please do.
Regards, Robert 

**PS -** The motherboard which is installed in the printer I have NOT disassembled. I don't know what "sits" in it. Is it the original TMC or are these clones ....

And by the way, one more question:
On my printer there is no way to print faster than 55 ~ 60mm/s and acceleration around 400 ~ 500.
When I try to set a higher speed, I end up with a massive shift of layers mainly in the Y-axis (X-axis also, but less frequently) and the printer makes such a scary noise. Some kind of growling and whirring ...
Is this "behaviour" normal at speeds of around 70~80mm/s?
And maybe you know what is the PROPER VREF for the motors in the X and Y axes ?

_PS2. translated by DeepL_

**Reply by 1coderookie** [reply]
> For a klipper for me it is too early (which I sincerely regret) because I am aware of my unfamiliarity with the subject.   

I thought the same when I started and so I used the stock fw for a few weeks at the beginning, just to get 'used' to everything. But frustration became too big and so I attempted using Klipper. Turned out it was *much* easier than I thought and after reading the docs and adjusting certain settings to my needs I regretted that I didn't do it right away.  
Really, even if you're an absolute beginner: go for it. Use the files from my [Klipper4KobraGoNeo repo](https://github.com/1coderookie/Klipper4KobraGoNeo), so you don't have to build the bin file and have a printer.cfg for the Neo you can start with. Read the docs and tune the settings to your needs and you're good. If you somehow can't manage to get comfortable with it, you can always go back to stock fw. But it's really worth the effort..  

> Is it worth "playing" with such modifications?  

*Imho*: no, as you'd have to solder them in place. Assuming you're not used to solder SMD parts with a heatgun, I wouldn't go for it as *I personally* don't see a real benefit here to do so. But I don't really know anything about these GC6609..  
Rather than messing around with SMD soldering here, I'd then buy a mainboard which can already be equipped with TMC2209, BTT has many models to offer. They're not even that expensive. But also here: *I personally* don't see a real benefit to do so as well, especially as your mobo is working fine and you now even have a spare board (where you'd have to glue the heatsinks back in place then of course..) which you could just swap in without changing *anything* if the first one somehow gets defective.  

> The motherboard which is installed in the printer I have NOT disassembled. I don't know what "sits" in it. Is it the original TMC or are these clones ....  

Does it have silver aluminum heatsinks as well? Or black ones? If silver, then I'd say it's most likely the case that those are GC6609 as well, as the silver heatsinks might be the indicator for that. At least at the mobos that have been inspected&reported all of the ones which were using the GC6609 had silver heatsinks.   

> When I try to set a higher speed, I end up with a massive shift of layers mainly in the Y-axis (X-axis also, but less frequently) and the printer makes such a scary noise. Some kind of growling and whirring ...  

Dunno if that sound comes from skipped steps tbh, but I personally would check the tension of the wheels and the tension of the belts first of all. Users often tend to run them too tight which can also cause issues and skipped steps.   
Besides that, try to identify the exact source of the sound. 

Just as a sidenote: also have an eye on the printing temp. The faster you wanna print, the hotter you need to print as well. So let's assume you're printing at the lower end of the temprange of your filament at 60mm/s, it *might* happen that it's too cold when trying to print at 80mm/s and therefore the higher flowrate which is needed can't be reached. To compensate that, you'd have to raise the printing temperature a bit when increading the printing speed. 

> And maybe you know what is the PROPER VREF for the motors in the X and Y axes ?   

The Vref is the proper one which is high enough that no step skipping occurs but which is low enough that they don't get too hot.  
I can't give you values, but I personally never had issues with the Vrefs that were set, I never changed anything here.

**Comment by robotm-team** [comment]
Thank you for your factual and comprehensive reply.
I've picked up a Wyse 3040 - I'll see when (as I'm rather sure it will) this clipper beats me. 
Regards.
PS. Translated by DeepL

**Comment by jokubasver** [comment]
I don't have too much time for a full write-up right now, but I will write a quick summary: My Kobra Neo, which was purchased in early January of 2023, uses original TMC2208 drivers, and I can confirm that the UART pins are not connected to the MCU - hence why I was unable to control the drivers using Marlin. So it seems that motherboards with GC6609 do connect UART to the MCU then? 

Turns out Klipper does support software serial, and even better - a single wire and any GPIO pin can be used for TX/RX for TMC drivers. So I wired up the UART pins of the four TMC2208 drivers to the free and available pins used in the LCD connector, and was able to implement TMC control via UART using the stock MB and Klipper: 
![image](https://github.com/1coderookie/KobraGoNeoInsights/assets/4341386/518f0af3-7b1e-4789-a4c8-029e63e0a022)

Once I have time, I will write a full write-up on how to do this.

**Reply by AshPugczup** [reply]
can you share any details how you did that any electric schematic or a photos?

**Comment by vfiuchcikicshuusrch** [comment]
Hi!
I decided to dig into the question of connecting the GC6609 via UART on my Kobra Go with trigorilla 3.0.7. And here’s what I found while studying the topic...
In the original Marlin firmware there are lines related to connecting the 6609 via UART and configuring the stepper driver addresses.
All the drivers on the board are connected to a single processor pin PB2. This is also confirmed by the code in the original firmware.

```
#if 1//HAS_TMC_UART
  /**
   * TMC2208/TMC2209 stepper drivers
   *
   * Hardware serial communication ports.
   * If undefined software serial is used according to the pins below
   */
#define TMC6609

//#define X_HARDWARE_SERIAL  Serial1
//#define Y_HARDWARE_SERIAL  Serial1
//#define Z_HARDWARE_SERIAL  Serial1
//#define E0_HARDWARE_SERIAL Serial1

//#define E1_HARDWARE_SERIAL Serial3

//
// Software serial
//
  #define X_SERIAL_TX_PIN                   PB2
  #define X_SERIAL_RX_PIN                   PB2

  #define Y_SERIAL_TX_PIN                   X_SERIAL_TX_PIN
  #define Y_SERIAL_RX_PIN                   X_SERIAL_RX_PIN

  #define Z_SERIAL_TX_PIN                   X_SERIAL_TX_PIN
  #define Z_SERIAL_RX_PIN                   X_SERIAL_RX_PIN

  #define E0_SERIAL_TX_PIN                  X_SERIAL_TX_PIN
  #define E0_SERIAL_RX_PIN                  X_SERIAL_RX_PIN

  // Reduce baud rate to improve software serial reliability
  #define TMC_BAUD_RATE                     19200
#endif
```

```
#define  X_SLAVE_ADDRESS 0
#define  Y_SLAVE_ADDRESS 1
#define  Z_SLAVE_ADDRESS 2
#define E0_SLAVE_ADDRESS 3
```

The M122 command executed in Marlin shows the following — apparently the drivers are indeed working in UART mode.

```
14:21:05.076 : N18 M122*27
14:21:05.079 : X	Y	Z	E
14:21:05.098 : Enabled		false	false	false	false
14:21:05.100 : Set current	800	800	800	600
14:21:05.118 : RMS current	795	795	795	581
14:21:05.143 : MAX current	1121	1121	1121	819
14:21:05.146 : Run current	25/31	25/31	25/31	18/31
14:21:05.149 : Hold current	12/31	12/31	12/31	9/31
14:21:05.167 : CS actual	18/31	18/31	18/31	18/31
14:21:05.167 : PWM scale
14:21:05.185 : vsense		1=.18	1=.18	1=.18	1=.18
14:21:05.210 : stealthChop	true	false	true	true
14:21:05.246 : msteps		16	16	16	16
14:21:05.272 : interp		true	true	true	true
14:21:05.289 : tstep		max	max	max	max
14:21:05.307 : PWM thresh.
14:21:05.307 : [mm/s]
14:21:05.332 : OT prewarn	false	false	false	false
14:21:05.378 : pwm scale sum	21	0	21	21
14:21:05.420 : pwm scale auto	0	0	0	0
14:21:05.459 : pwm offset auto	36	36	36	0
14:21:05.483 : pwm grad auto	14	14	14	14
14:21:05.538 : off time	4	4	0	4
14:21:05.556 : blank time	24	24	24	24
14:21:05.556 : hysteresis
14:21:05.597 : -end		2	-3	2	-3
14:21:05.627 : -start		1	1	1	1
14:21:05.627 : Stallguard thrs
14:21:05.670 : uStep count	0	8	8	8
14:21:05.670 : DRVSTATUS	X	Y	Z	E
14:21:05.672 : sg_result
14:21:05.700 : stst
14:21:05.717 : olb
14:21:05.753 : ola
14:21:05.768 : s2gb
14:21:05.784 : s2ga
14:21:05.801 : otpw
14:21:05.816 : ot
14:21:05.832 : 157C
14:21:05.873 : 150C
14:21:05.905 : 143C
14:21:05.952 : 120C
14:21:05.952 : s2vsa
14:21:05.952 : s2vsb
14:21:05.954 : Driver registers:
14:21:05.960 : X	0xC0:12:00:00
14:21:05.965 : Y	0xC0:12:00:00
14:21:05.969 : Z	0xC0:12:00:00
14:21:05.974 : E	0xC0:12:00:00
14:21:05.987 : Testing X connection... OK
14:21:06.002 : Testing Y connection... OK
14:21:06.007 : Testing Z connection... OK
14:21:06.011 : Testing E connection... OK




14:22:22.534 : N19 M122*26
14:22:22.538 : X	Y	Z	E
14:22:22.555 : Enabled		false	false	false	false
14:22:22.558 : Set current	800	800	800	600
14:22:22.599 : RMS current	1436	795	1436	581
14:22:22.617 : MAX current	1121	1121	1121	819
14:22:22.620 : Run current	25/31	25/31	25/31	18/31
14:22:22.624 : Hold current	12/31	12/31	12/31	9/31
14:22:22.641 : CS actual	18/31	18/31	18/31	18/31
14:22:22.641 : PWM scale
14:22:22.667 : vsense		0=.325	1=.18	1=.18	1=.18
14:22:22.692 : stealthChop	true	true	false	true
14:22:22.708 : msteps		16	16	16	16
14:22:22.738 : interp		true	true	false	true
14:22:22.787 : tstep		max	0	0	0
14:22:22.837 : PWM thresh.
14:22:22.837 : [mm/s]
14:22:22.867 : OT prewarn	false	false	false	false
14:22:22.921 : pwm scale sum	0	0	0	21
14:22:22.951 : pwm scale auto	0	0	0	0
14:22:22.985 : pwm offset auto	36	36	36	36
14:22:23.021 : pwm grad auto	14	14	0	14
14:22:23.046 : off time	4	4	4	4
14:22:23.092 : blank time	24	24	16	24
14:22:23.092 : hysteresis
14:22:23.124 : -end		2	2	2	2
14:22:23.143 : -start		1	1	1	1
14:22:23.143 : Stallguard thrs
14:22:23.178 : uStep count	8	8	8	8
14:22:23.178 : DRVSTATUS	X	Y	Z	E
14:22:23.180 : sg_result
14:22:23.216 : stst			*
14:22:23.232 : olb
14:22:23.248 : ola
14:22:23.264 : s2gb
14:22:23.301 : s2ga
14:22:23.317 : otpw
14:22:23.332 : ot
14:22:23.361 : 157C
14:22:23.389 : 150C
14:22:23.413 : 143C
14:22:23.435 : 120C
14:22:23.435 : s2vsa
14:22:23.435 : s2vsb
14:22:23.436 : Driver registers:
14:22:23.442 : X	0xC0:12:00:00
14:22:23.447 : Y	0xC0:12:00:00
14:22:23.452 : Z	0xC0:12:00:00
14:22:23.458 : E	0xC0:12:00:00
14:22:23.467 : Testing X connection... OK
14:22:23.473 : Testing Y connection... OK
14:22:23.479 : Testing Z connection... OK
14:22:23.484 : Testing E connection... OK



14:23:17.647 : N20 M122*16
14:23:17.651 : X	Y	Z	E
14:23:17.676 : Enabled		false	false	false	false
14:23:17.679 : Set current	800	800	800	600
14:23:17.696 : RMS current	795	795	795	581
14:23:17.723 : MAX current	1121	1121	1121	819
14:23:17.726 : Run current	25/31	25/31	25/31	18/31
14:23:17.729 : Hold current	12/31	12/31	12/31	9/31
14:23:17.755 : CS actual	18/31	18/31	18/31	18/31
14:23:17.755 : PWM scale
14:23:17.797 : vsense		1=.18	1=.18	1=.18	1=.18
14:23:17.822 : stealthChop	true	true	true	false
14:23:17.855 : msteps		16	16	16	16
14:23:17.875 : interp		true	true	true	true
14:23:17.914 : tstep		max	0	max	max
14:23:17.947 : PWM thresh.
14:23:17.947 : [mm/s]
14:23:17.986 : OT prewarn	false	false	false	false
14:23:18.011 : pwm scale sum	21	21	21	21
14:23:18.033 : pwm scale auto	0	0	0	0
14:23:18.062 : pwm offset auto	0	36	36	36
14:23:18.087 : pwm grad auto	14	14	14	14
14:23:18.123 : off time	4	4	4	0
14:23:18.151 : blank time	24	24	24	24
14:23:18.151 : hysteresis
14:23:18.168 : -end		2	2	2	2
14:23:18.193 : -start		1	1	1	1
14:23:18.193 : Stallguard thrs
14:23:18.212 : uStep count	8	8	8	8
14:23:18.213 : DRVSTATUS	X	Y	Z	E
14:23:18.214 : sg_result
14:23:18.247 : stst					*
14:23:18.275 : olb
14:23:18.299 : ola
14:23:18.331 : s2gb
14:23:18.355 : s2ga
14:23:18.387 : otpw
14:23:18.412 : ot
14:23:18.436 : 157C
14:23:18.468 : 150C
14:23:18.484 : 143C
14:23:18.513 : 120C
14:23:18.513 : s2vsa
14:23:18.513 : s2vsb
14:23:18.515 : Driver registers:
14:23:18.521 : X	0xC0:12:00:00
14:23:18.527 : Y	0xC0:12:00:00
14:23:18.541 : Z	0xC0:12:00:00
14:23:18.554 : E	0xC0:12:00:00
14:23:18.563 : Testing X connection... OK
14:23:18.569 : Testing Y connection... OK
14:23:18.576 : Testing Z connection... OK
14:23:18.581 : Testing E connection... OK
```

However, when configuring Klipper with these parameters, only one driver responds over UART — and even that one not fully...

```
Unable to read tmc uart 'extruder' register GCONF
========== Queried registers ==========
SGTHRS:     00000000
TPOWERDOWN: 00000014 tpowerdown=20
COOLCONF:   00000000
TCOOLTHRS:  00000000
TPWMTHRS:   000fffff tpwmthrs=1048575
IHOLD_IRUN: 00080f0f ihold=15 irun=15 iholddelay=8
SLAVECONF:  00000200 senddelay=2
========== Write-only registers ==========
dump_tmc stepper=extruder

Unable to read tmc uart 'stepper_z' register GCONF
========== Queried registers ==========
SGTHRS:     00000000
TPOWERDOWN: 00000014 tpowerdown=20
COOLCONF:   00000000
TCOOLTHRS:  00000000
TPWMTHRS:   00000013 tpwmthrs=19
IHOLD_IRUN: 00081313 ihold=19 irun=19 iholddelay=8
SLAVECONF:  00000200 senddelay=2
========== Write-only registers ==========
dump_tmc stepper=stepper_z

Unable to read tmc uart 'stepper_y' register GCONF
========== Queried registers ==========
SGTHRS:     00000000
TPOWERDOWN: 00000014 tpowerdown=20
COOLCONF:   00000000
TCOOLTHRS:  00000000
TPWMTHRS:   0000002f tpwmthrs=47
IHOLD_IRUN: 00081313 ihold=19 irun=19 iholddelay=8
SLAVECONF:  00000200 senddelay=2
========== Write-only registers ==========
dump_tmc stepper=stepper_y

Unable to read tmc uart 'stepper_x' register OTP_READ
IFCNT:      00000050 ifcnt=80
GSTAT:      00000001 reset=1(Reset)
GCONF:      000001c0 pdn_disable=1 mstep_reg_select=1 multistep_filt=1
========== Queried registers ==========
SGTHRS:     00000000
TPOWERDOWN: 00000014 tpowerdown=20
COOLCONF:   00000000
TCOOLTHRS:  00000000
TPWMTHRS:   0000002f tpwmthrs=47
IHOLD_IRUN: 00081313 ihold=19 irun=19 iholddelay=8
SLAVECONF:  00000200 senddelay=2
========== Write-only registers ==========
dump_tmc stepper=stepper_x
```

Any comments are welcome!

Maybe sombody have high quality photo drivers places ? ms1 ms2 are different or the same on all drivers ?
Marlin worked with uart ? it's true for this printer ?
